checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -0,0 +1,834 @@
+    // the values are the set fields on which those methods are called
+    Map<String, Set<String>> methodToFields = new LinkedHashMap<>();
+    for (VariableElement releasedField : releasedFields) {
+      List<String> mustCallValues = typeFactory.getMustCallValue(releasedField);
+      assert !mustCallValues.isEmpty()
+          : "Must-call obligation of owning field " + releasedField + " is deleted.";
+      assert mustCallValues.size() == 1
+          : "The must-call set ("
+              + mustCallValues
+              + ") of "
+              + releasedField
+              + "should be a singleton";
+      // The assumption is that the must-call set has only one element
+      String mustCallValue = mustCallValues.get(0);
+      String fieldName = "this." + releasedField.getSimpleName().toString();
+
+      methodToFields.computeIfAbsent(mustCallValue, k -> new HashSet<>()).add(fieldName);
+    }
+
+    for (String mustCallValue : methodToFields.keySet()) {
+      AnnotationMirror am =
+          createEnsuresCalledMethods(
+              methodToFields.get(mustCallValue).toArray(new String[0]),
+              new String[] {mustCallValue});
+      WholeProgramInference wpi = typeFactory.getWholeProgramInference();
+      wpi.addMethodDeclarationAnnotation(methodElt, am);
+    }
+  }
+
+  /**
+   * Possibly adds an InheritableMustCall annotation on the enclosing class. If the class already
+   * has a non-empty MustCall type (which is inherited from one of its superclasses), this method
+   * does nothing, in order to avoid infinite iteration. Otherwise, if the current method is not
+   * private and satisfies the must-call obligations of all the owning fields, it adds an
+   * InheritableMustCall annotation to the enclosing class.
+   */
+  private void addOrUpdateMustCall() {
+    ClassTree classTree = TreePathUtil.enclosingClass(typeFactory.getPath(methodTree));
+    TypeElement typeElement = TreeUtils.elementFromDeclaration(classTree);
+    if (typeElement == null) {
+      return;
+    }
+
+    WholeProgramInference wpi = typeFactory.getWholeProgramInference();
+    List<String> currentMustCallValues = typeFactory.getMustCallValue(typeElement);
+    if (!currentMustCallValues.isEmpty()) {
+      // If the class already has a MustCall annotation which is inherited from a superclass,
+      // do nothing.
+      if (typeElement.getSuperclass() != null) {
+        TypeMirror superType = typeElement.getSuperclass();
+        TypeElement superTypeElement = TypesUtils.getTypeElement(superType);
+        if (superTypeElement != null && !typeFactory.getMustCallValue(superTypeElement).isEmpty()) {
+          return;
+        }
+      }
+      // Add the current @MustCall annotation to guarantee the termination property. This is
+      // necessary to avoid overwriting @MustCall annotations on the class declaration when there
+      // are multiple methods that could fulfill the must-call obligation.
```

Comment:
This part of the code simply adds the existing @MustCall annotation instead of recomputing it. It is helpful to avoid flip-flopping between different iterations when there are multiple finalizers in the class.